media/common/src/main/java/io/helidon/media/common/DataChunkInputStream.java

```diff
@@ -132,32 +132,38 @@ public int read(byte[] buf, int off, int len) throws IOException {
                 return -1;
             }
 
-            ByteBuffer currentBuffer = chunk.data();
-
-            if (currentBuffer.position() == 0) {
-                LOGGER.finest(() -> "Reading chunk ID: " + chunk.id());
-            }
-
-            // If there is anything to read, then read as much as fits into buf
-            int rem = currentBuffer.remaining();
-            if (len > rem) {
-                len = rem;
+            ByteBuffer[] currentBuffers = chunk.data();
+            int count = 0;
+            for (int i = 0; i < currentBuffers.length; i++) {
+                if (i == 0 && currentBuffers[i].position() == 0) {
+                    LOGGER.finest(() -> "Reading chunk ID: " + chunk.id());
+                }
+                // If there is anything to read, then read as much as fits into buf
+                int rem = currentBuffers[i].remaining();
+                if (rem == 0) {
+                    continue;
+                }
+                int currentLen = len;
+                if (currentLen > rem) {
+                    currentLen = rem;
+                }
+                currentBuffers[i].get(buf, off, currentLen);
+                off += currentLen;
+                count += currentLen;
+
+                // Chunk is consumed entirely - release the chunk, and prefetch a new chunk; do not
+                // wait for it to arrive - the next read may have to wait less.
+                //
+                // Assert: it is safe to request new chunks eagerly - there is no mechanism
+                // to push back unconsumed data, so we can assume we own all the chunks,
+                // consumed and unconsumed.
+                if (i == currentBuffers.length - 1 && currentLen == rem) {
```

Comment:
What would happen with this condition if the last `ByteBuffer` in the chunk returns 0 and continues in line 144?

Reply:
Good point. I will make another pass at this and add a unit test.

Reply:
@spericas, I've made changes, please take a look and let me know.

Reply:
@romain-grecourt I'll take a closer look. At first glance, the code is still a bit odd; it seems that the `DataChunk` abstraction needs to be improved so we can call `get(byte[], int off, int len)` and so that it can iterate over its `ByteBuffer`'s, copy the data and also remember which buffer it is reading from. 
Your code is always starting from the first buffer on each call to `read(...)`, even though buffers may have been consumed in an earlier call. If the `DataChunk` abstraction is enhanced, then`DataChunkInputStream` shouldn't need to change too much.